### PR TITLE
Methods to easily access Authorization information

### PR DESCRIPTION
**Pull request by cimnine**
## Summary

This PR adds information about the HTTP request principal (extracted from either the Basic auth username or from a claim of the JWT token presented in Bearer token auth).

Besides, the PR adds this:

- Improved configuration sanity checks:
  - Only empty values are substituted with default values.
  - An error is thrown, when a value is given, but it can't be parsed (e.g. parsing ints or durations).
  - The error messages have been improved as well
- In some functions, cohesive functionality has been extracted into a separate function

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Categorize the PR by setting a good title and adding one of the labels:
  `bug`, `enhancement`, `documentation`, `change`, `breaking`,
  as they show up in the changelog
- [x] Update the documentation.
- [x] Update tests.
- [ ] ~~Link this PR to related issues.~~

### CODE REVIEW

File: docs/modules/ROOT/pages/index.adoc [hunk 1]
```diff
@@ -14,30 +14,36 @@ NOTE: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOUL
...
 |_MUST be provided, no default value is defined._
 |https://www.random.org/strings/?num=2&len=20&digits=on&upperalpha=on&loweralpha=on&unique=on&format=plain&rnd=new[Random String]
 
-|`OSB_NAMESPACE`
-|Which namespace the service broker shall use to create the resources.
-|_MUST be provided, no default_
-|`crossplane-service-broker`
+|`OSB_USERNAME_CLAIM`
+|If a https://tools.ietf.org/html/rfc7519[JWT] https://tools.ietf.org/html/rfc6750[`Bearer` token] is presented,
+then the value of this variable defines the claim that is considered to contain the username of the request principal.
```
Comment (ccremer) [inline docs/modules/ROOT/pages/index.adoc:41]: should be one line, even if it goes beyond the 80-char screen ;)

File: pkg/api/auth/principal.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/pascaldekloe/jwt"
+	"github.com/vshn/crossplane-service-broker/pkg/config"
+)
+
+// Principal represents the entity (person or system) in who's name the current request is made.
+// It does not represent the _origin_ of the request.
+// For example, a system might send a request in the name of a person.
+type Principal string
+
+// PrincipalFromContext extracts the Principal on who's name the given http.Request was made based on the information
```
Comment (ccremer) [inline pkg/api/auth/principal.go:16]:
```suggestion
// PrincipalFromContext extracts the Principal on whose name the given http.Request was made based on the information
```

File: pkg/api/auth/principal.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	case authenticationMethodStr == AuthenticationMethodBasicAuth:
+		return getPrincipalFromBasicAuth(ctx)
+	case authenticationMethodStr == AuthenticationMethodBearerToken:
+		return getPrincipalFromBearerToken(ctx, cfg)
+	default:
+		return "", fmt.Errorf("unknown authorication format '%s'", authenticationMethodStr)
+	}
+}
+
+// getPrincipalFromBearerToken extracts the Principal on who's name the given http.Request was made based on the
```
Comment (ccremer) [inline pkg/api/auth/principal.go:39]:
```suggestion
// getPrincipalFromBearerToken extracts the Principal on whose name the given http.Request was made based on the
```

File: pkg/api/auth/principal.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+	username, ok := claim.String(cfg.UsernameClaim)
+	if !ok {
+		return "", fmt.Errorf("principal token contains no claim named '%s'", cfg.UsernameClaim)
+	}
+
+	return Principal(username), nil
+}
+
+// getPrincipalFromBasicAuth extracts the Principal on who's name the given http.Request was made based on the
```
Comment (ccremer) [inline pkg/api/auth/principal.go:60]:
```suggestion
// getPrincipalFromBasicAuth extracts the Principal on whose name the given http.Request was made based on the
```

File: pkg/config/config_test.go [hunk 5]
```diff
@@ -14,50 +13,86 @@ func Test_ReadConfig(t *testing.T) {
...
 		"password required": {
 			env: map[string]string{
-				"OSB_SERVICE_IDS": "1,2,3",
-				"OSB_USERNAME":    "user",
+				EnvServiceIDs: "1,2,3",
+				EnvUsername:   "user",
 			},
 			config: nil,
-			err:    "OSB_PASSWORD is required",
+			err:    "OSB_PASSWORD is required, but was not defined or is empty",
```
Comment (ccremer) [inline pkg/config/config_test.go:47]: Maybe use the variable here as well?
```suggestion
			err:    EnvPassword + " is required, but was not defined or is empty",
```

File: docs/modules/ROOT/pages/index.adoc (hunk 1 above)
Comment (cimnine) [inline docs/modules/ROOT/pages/index.adoc:41]:
```suggestion
|If a https://tools.ietf.org/html/rfc7519[JWT] https://tools.ietf.org/html/rfc6750[`Bearer` token] is presented, then the value of this variable defines the claim that is considered to contain the username of the request principal.
```

**Review by ccremer — APPROVED** [review]

**Pull request merged**